airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/DebeziumRecordPublisher.java

```diff
@@ -0,0 +1,161 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.source.postgres;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.integrations.source.jdbc.AirbyteFileOffsetBackingStore;
+import io.airbyte.protocol.models.AirbyteStream;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.ConfiguredAirbyteStream;
+import io.debezium.engine.ChangeEvent;
+import io.debezium.engine.DebeziumEngine;
+import io.debezium.engine.format.Json;
+import io.debezium.engine.spi.OffsetCommitPolicy.AlwaysCommitOffsetPolicy;
+import java.util.Properties;
+import java.util.Queue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.codehaus.plexus.util.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DebeziumRecordPublisher implements AutoCloseable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DebeziumRecordPublisher.class);
+  private final ExecutorService executor;
+  private DebeziumEngine<ChangeEvent<String, String>> engine;
+
+  private final JsonNode config;
+  private final ConfiguredAirbyteCatalog catalog;
+
+  private final AtomicBoolean hasClosed;
+  private final AtomicReference<Throwable> thrownError;
+
+  public DebeziumRecordPublisher(JsonNode config, ConfiguredAirbyteCatalog catalog) {
+    this.config = config;
+    this.catalog = catalog;
+    this.hasClosed = new AtomicBoolean(false);
+    this.thrownError = new AtomicReference<>();
+    this.executor = Executors.newSingleThreadExecutor();
+  }
+
+  public void start(Queue<ChangeEvent<String, String>> queue) {
+    engine = DebeziumEngine.create(Json.class)
+        .using(getDebeziumProperties(config, catalog))
+        .using(new AlwaysCommitOffsetPolicy())
+        .notifying(queue::add)
+        .using((success, message, error) -> {
+          LOGGER.info("completed!");
+          thrownError.set(error);
+        })
+        .build();
+
+    // Run the engine asynchronously ...
+    executor.execute(engine);
+  }
+
+  public boolean hasClosed() {
+    return hasClosed.get();
+  }
+
+  public synchronized void close() throws Exception {
```

Comment:
why dont you make close idempotent instead of synchronized?
```
public void close() throws Exception {
  if (hasclosed.compareAndSet(false, true) {
     close stuff & shutdown
  }
}
```

Reply:
that doesn't quite give us the guarantee that we want. there are things we care about here:
1. we only want the internals of close to be called no more than once
2. we only want to set `isClosed` to `true` _after_ the internals of close have run.

i think what you're describing only gives us 1. if we introduced a second boolean `isClosing` we might be able to use this technique but it leads to a weird behavior where if close is called twice, the second call can return from close while the object still isn't actually closed (unless we add a busy wait). gets kinda complicated. lmk if you think i'm missing something. definitely happy to try a different approach.

Reply:
Nope, you're correct!